fix(datepicker): reset highlight on value change

minseoksuh opened the pull request:
https://github.com/uber/baseweb/issues/3067

#### Description

- added resetHighlightOnValueChange props for calender in datepicker/types.js
- added logic to reset highlight in componentDidUpdate in calendar.js
- wrote unit test for resetHighlihgtOnValue in calendar.test.js
- added description of the feature in documentation

#### Scope

- [x] Patch: Bug Fix
- [ ] Minor: New Feature
- [ ] Major: Breaking Change

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/uber/baseweb?pullRequest=3147) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/uber/baseweb?pullRequest=3147) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/uber/baseweb?pullRequest=3147) it.</sub>

gergelyke commented:
@chasestarr @minseoksuh I wonder if this should be the default behavior? that case, we wouldn't even have to introduce a new prop

chasestarr commented:
Yes, I agree that this can be the default

gergelyke commented:
@minseoksuh would you mind updating the PR?

nadiia commented:
A few things here:
- I couldn't reproduce the issue you described in the ticket on the latest version
- The codesandbox example in [#3067](https://github.com/uber/baseweb/issues/3067#issuecomment-601001375) doesn't demonstrate the issue described, also, it doesn't use the latest baseui version. Could you provide the repro of the issue with the latest version? 
- We should not add a new component property for this issue

minseoksuh commented:
> A few things here:
> 
> * The codesandbox example in [#3067](https://github.com/uber/baseweb/issues/3067#issuecomment-601001375) doesn't demonstrate the issue described, also, it doesn't use the latest baseui version. But I could repro the issue on the masters storybook
> * We should not add a new component property for this issue
> * We should set the highlighted value in the state to the first selected value if any, otherwise to the highlighted value that comes from props, if none is provided and no selected value then we set it to today's date.

1. I just added the quickSelect option to the calendar, and I updated the sandbox to use the latest version. When I did this, the issue action was not reproduced. However, in this case, today was highlighted even though the selected range have changed. (Different response from the original issue, still looks glitchy)

2. Ok I can get rid of the props.

3. I think it makes sense to have the first date of the currently selectedDate highlighted. Should I implement @nadiia 's suggestion and re-commit to this PR?

Thanks

minseoksuh commented on src/datepicker/calendar.js:
```diff
@@ -112,8 +111,16 @@ export default class Calendar extends React.Component<
     }
 
-    if (prevProps.value !== this.props.value) {
+    if (prevProps.value !== this.props.value && !this.state.focused) {
+      if (this.props.value instanceof Date) {
+        this.setState({highlightedDate: this.props.value});
+      } else if (this.props.value instanceof Array) {
+        this.setState({highlightedDate: this.props.value[0]});
+      } else {
+        this.setState({highlightedDate: new Date()});
+      }
```
Changed this part to incorporate Nadia's suggestion. Added !this.state.focused to the conditional because without it highlight on click (specifically clicking to select range) is bugged. The second click to select an end day of range will render the endDate as blank (without the number inside rendered). !this.state.focused condition makes sure the reset occurs only when value is changed through external measures (like quickselect)

minseoksuh commented on src/datepicker/calendar.js:
```diff
@@ -372,15 +379,8 @@ export default class Calendar extends React.Component<
     const selected = this.getSingleDate(value);
-    let nextState;
-    if (selected && isSameMonth(selected, date) && isSameYear(selected, date)) {
-      nextState = {highlightedDate: selected};
-    } else {
-      nextState = {
-        highlightedDate: date,
-      };
-    }
-    this.setState(nextState);
+    if (selected) this.setState({highlightedDate: selected});
+    else this.setState({highlightedDate: new Date()});
```
explained this part in the commit message.

nadiia commented:
Works well, and good to go after addressing the last couple comments 👍

minseoksuh commented:
> Works well, and good to go after addressing the last couple comments 👍

Yup.

Added the check logic and added comments for the update section.

Thanks for the feedback.

Please let me know if anything else needs to be done.

nadiia commented:
The issue is when keyboard navigating to a next/prev month there is no focusable date in the calendar in a new month rendered and therefore no way to tab into the calendar.
![datepicker-keyboard-nav](https://user-images.githubusercontent.com/1920399/79017951-6e23c380-7b27-11ea-9a71-42e421a12b00.gif)

minseoksuh commented:
> The issue is when keyboard navigating to a next/prev month there is no focusable date in the calendar in a new month rendered and therefore no way to tab into the calendar.

Allowed the user to tab into Sunday's of the calendar (if that is ok) even if it is not highlighted.
Also, when tabbing out of the calender's day section, the highlight remained, so added a logic to reset the highlight to today's date in that situation.

nadiia commented:
![2020-04-14 at 3 33 42](https://user-images.githubusercontent.com/1920399/79215784-f690b580-7e00-11ea-96b5-309329a5d66c.gif)

minseoksuh commented:
> It adds a regression in the navigation, seel below:
> ![2020-04-14 at 3 33 42](https://user-images.githubusercontent.com/1920399/79215784-f690b580-7e00-11ea-96b5-309329a5d66c.gif)

Uh don't understand the problem.

Can you please describe the proper behavior in more detail?

nadiia commented:
Sorry about it. When a calendar day is focused and the tab key is pressed it should focus on the next focusable element outside of the calendar month. So it would be the quick select input if any or the first focusable element (previous month button) in the calendar's header. In the current change tho it tabs through the days in the calendar before reaching anything outside of the calendar. I added an e2e test for it in #3207

minseoksuh commented:
> Sorry about it. When a calendar day is focused and the tab key is pressed it should focus on the next focusable element outside of the calendar month. So it would be the quick select input if any or the first focusable element (previous month button) in the calendar's header. In the current change tho it tabs through the days in the calendar before reaching anything outside of the calendar. I added an e2e test for it in #3207
> 
> Basically, only a single day at a time in a calendar month should have a tabindex=0. If there is no user-highlighted date it should be the 1st of the month.

Fixed the behavior. Thanks for the detailed explanation.

minseoksuh commented:
@nadiia Hi just wondering how the PR is going. Let me know if anything else need to be done.
Thanks!

nadiia commented:
Hi @minseoksuh, since this change affects too many details in the calendar navigation and, unfortunately, we didn't have enough tests to cover all those interesting and tricky behaviors, I assumed it would make sense to take care of it on our end. So I opened a PR (#3256) with a fix for the brought up issue. But I would like to thank you for starting working on the fix.
Please check out the built preview of the fix and I'll be closing this PR.
